designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,414 @@
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageB" />
+  </ItemGroup>
+
+</Project>
+```
+
+#### Gesture
+
+The customer updates `PackageA` to version `2.0.0` in Package Manager UI.
+
+#### After
+
+`$(repo_root)\Directory.Packages.props`
+
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="2.0.0" />
+    <PackageVersion Include="PackageB" Version="4.5.6" />
+  </ItemGroup>
+</Project>
+```
+
+`$(repo_root)\Project1\Project1.csproj`
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageA" />
+  </ItemGroup>
+
+</Project>
+```
+
+`$(repo_root)\Project2\Project2.csproj`
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageB" />
+  </ItemGroup>
+
+</Project>
+```
+
+## Requirements
+
+The requirements are defined in the [feature spec](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions).
+
+Note that:
+
+* It's valid for some, not all, projects in the solution to use centrally managed package versions.
+* It's valid for different projects using central package versions to use different `Directory.Packages.props` files.
+* As is valid now, it's valid for some projects to use `packages.config`, while others use `PackageReference`. The `packages.config` projects cannot use centrally managed package versions, and individual `PackageReference` projects may opt in or out of central package version management.
+
+## Solution
+
+A new service should be created. This service will have multiple responsibilities:
+
+* Write changes to `Directory.Packages.props` files to disk.
+* Find which `Directory.Packages.props` file needs to be changed given a package install for a specific project.
+* Orchestrate efficient updating and nomination of projects. In particular, when project files and `Directory.Packages.props` files need to change due to a single customer gesture, avoid multiple project nominations and multiple restores due to multiple file changes.
+
+This new service will by used by NuGet, but only for projects that have centrally managed versions enabled. Projects that do not use CPVM will continue to work as they do today.
+
+Proposal: When a project is using `CPVM`, NuGet will only use this new service and will no longer call the existing project system APIs
+
+Proposed API. Very subject to change as we start implementing and discover better ideas:
+
+```cs
+interface IDirectoryPackagesPropsService
+{
+  /// <summary>
+  /// Add a PackageReference to the projects provided. If the packageId does
+  /// </summary>
+  Task<bool> AddPackageAsync(List<string> projects, string packageId, string packageVersion);
```

Comment:
projects is like a list of projects' full paths?

Reply:
What should be the behavior on error? All or none of the projects updated?

Reply:
Can these operations fail because of user settings/behavior? If yes how should these types of errors bubbled up? 

Can these methods throw? If yes in what conditions?